### DISCUSSION TITLE
Invalid header after flashing firmware

### DISCUSSION THREAD

**Discussion by amigthea** [awawa-dev/HyperSPI#43] [Announcements]
after I flash the .bin and I try to connect to serial port it outputs continuosly this log:

```
st:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
invalid header: 0x28202b20
invalid header: 0x28202b20
invalid header: 0x28202b20
invalid header: 0x28202b20
invalid header: 0x28202b20
invalid header: 0x28202b20
invalid header: 0x28202b20
invalid header: 0x28202b20
```

it is expected?

**Comment by awawa-dev** [comment]
no
more details: which firmware, board model etc

**Comment by amigthea** [comment]
Platform: ESP32 MH-ET Live

I followed the [wiki](https://github.com/awawa-dev/HyperSPI/wiki) instructions:

- Forked
- Edited the platformio.ini (only the build_flags line)

```
[env:SK6812_RGBW_NEUTRAL]
build_flags = -DNEOPIXEL_RGBW -DDATA_PIN=2 ${env.build_flags} -DSECOND_SEGMENT_START_INDEX=112 -DSECOND_SEGMENT_DATA_PIN=4 -DSECOND_SEGMENT_REVERSED
custom_prog_version = esp32_SK6812_RGBW_NEUTRAL

board = esp32dev
platform = ${esp32.platform}
lib_deps = ${esp32.lib_deps}
test_ignore = ${esp32.test_ignore}
```

- flashed the firmware

`./esptool write_flash 0x0 firmware_esp32_SK6812_RGBW_NEUTRAL.bin`

- it gives those errors

**Comment by awawa-dev** [comment]
You flashed the firmware at wrong address and damaged the bootloader. Use factory firmware:
https://github.com/awawa-dev/HyperSPI?tab=readme-ov-file#flashing-the-firmware

**Comment by amigthea** [comment]
ok, now I flashed the factory firmware and then the custom one at the correct address

```
./esptool write_flash 0x0 firmware_esp32_SK6812_RGBW_NEUTRAL.factory.bin
./esptool write_flash 0x10000 firmware_esp32_SK6812_RGBW_NEUTRAL.bin
```

and now it does not gives me those error anymore

but even if I configured the Led Hardware section as [shown by the guide](https://user-images.githubusercontent.com/69086569/193319124-0054f367-3d30-4e50-8c52-3683c7bbc50e.png) I get

`[LEDDEVICE0_AWA_SPI] Device 'awa_spi' is disabled due to an error: 'Failed to open device (/dev/spidev0.0). Error message: No such file or directory'`

tried to change that to `/dev/ttyACM0` to no avail, what could be the cause?

edit: the controller is usb connected to the rpi4 where HyperHDR is running

**Comment by awawa-dev** [comment]
/dev/spidev0.0 must be activated in the system first, usually HyperHDR installer tries to enable it but it requires a restart. Use raspi-config alternatively.

**Comment by amigthea** [comment]
thank you, that solved the Led Hardware recognition problem <3

I have the Video Capturing configuration up and running (confirmed by the Live Video preview) but both led segments are off and they do not turn on, and I'm trying to solve this. From a soldering perspective I have [this exact scheme](https://user-images.githubusercontent.com/85223482/222923979-f344349a-1f8b-4195-94ca-51721923359e.png) from the controller to the ledstrip and the wiring [show here](https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEglZwPUxmp0INSnMRpT0VLdCbmPrIFdAgoveADiBnUCZ39S8wz1m1CH9OVgL1oTIgT8haXcinDrjuLlb6gvMafV4sIk2b-8fPCTx1bz6Wgx2rENnVSW7lv3S3C5qWYaLO7cBhiH4jcYur73oT0T9q2774OMoimY9slYGCwDl3Z_xFdqjb5Z9WNaGHiPKQ/s6521/single_segment.png) from the controller to the rpi4, is that missing something, maybe another wire to the raspberry while using two segments?

If the wiring is ok, what could keep the leds off?

**Comment by awawa-dev** [comment]
Make sure that there is communication first (sometimes users use very strange or long wiring to connect Rpi to ESP SPI or they forget about ground which is **mandatory and as important as other SPI cables**) https://github.com/awawa-dev/HyperSPI?tab=readme-ov-file#performancedebug-output
run screen => run HyperHDR and some effect using configured SPI devices => stop HyperHDR => screen should show some statistics

both segments must be equal and total number of LEDs in HyperHDR = 2 * size of the segment

**Reply by amigthea** [reply]
> run screen => run HyperHDR and some effect using configured SPI devices => stop HyperHDR => screen should show some statistics

I tried this but I got no meaningful output, what could be the cause?

**Reply by awawa-dev** [reply]
I can't help with something like `no meaningful` because I have completely no idea what it means. On the main page you have described the connections and even a few photos showing what ground is involved in SPI (between GPIO Rpi and esp32) and in the next post you write about the ground for the level shifter and LEDs (equally important but this is the second circuit not involved in SPI communication). As for the cables, I have absolutely no idea whether such long cables will work because I recommended shorter ones: I guess it will be known after your tests.

**Reply by awawa-dev** [reply]
![obraz](https://github.com/awawa-dev/HyperSPI/assets/69086569/ca54416d-0f23-45eb-9623-8ba0e75f7f65)

**Reply by awawa-dev** [reply]
![obraz](https://github.com/awawa-dev/HyperSPI/assets/69086569/66091398-516b-4828-a51b-59a09d325bc5)

**Reply by amigthea** [reply]
first of all, I wanna thank you for the patience you're showing to my non-optimal interaction, I will be more specific in my communication.

After some trial, I triple checked everything and I think the problem resides in the controller itself; maybe it's the wrong model? I bought [this one](https://it.aliexpress.com/item/32964909906.html?spm=a2g0o.order_list.order_list_main.45.26d73696d6LcqM&gatewayAdapt=glo2ita) and `esptool read_mac` gives me

```
Detecting chip type... ESP32
Chip is ESP32-D0WD-V3 (revision v3.1)
```

ps: I already bought another batch, this time I ordered [the model you suggested](https://it.aliexpress.com/item/32836372640.html?spm=a2g0o.order_list.order_list_main.5.26d73696d6LcqM&gatewayAdapt=glo2ita) in another discussion.

Could this be the problem?

**Reply by amigthea** [reply]
update: visually inspecting the board I found the chip type ingraved on a little black chip, it's the CH9102F, is that chip usable in this project or I need the CH9102X (that I bought meanwhile)?

**Reply by awawa-dev** [reply]
The serial chip is not important in the HyperSPI project because we only use it for diagnostics, taking advantage of the low speed that any such chip can handle. You must have a correct SPI connection (including ground) and `screen` must confirm SPI communication and that frames are reaching from RPi to ESP32. You can test it without LEDs connected.

**Reply by amigthea** [reply]
thank you so much! I was able to pinpoint the problem, something went wrong when forking the project/creating the custom firmware. Now everything works, almost perfectly.

I have one last problem with the video capture source: even in the blackest scenes, the leds emit a dim light and do not completely turn off.
- If I select a fixed color source rgb(0,0,0) they turn off correctly.
- HDR to SDR tone mapping is disabled (but I should enabled it because I get [this](https://github.com/awawa-dev/HyperHDR/discussions/729#discussioncomment-8126042) problem with DV content)

What could cause that?

[this](https://imgur.com/a/jFIqe1v) is a screenshot from the video grabber, o a completely black scene, taken from Live video menu.

**Reply by awawa-dev** [reply]
set minimal backlight to 0 in the processing tab

**Comment by amigthea** [comment]
is 25-30cm considered a long wiring?
I have the ground from one of the segments all way down to the rpi, passing through the level shifter and the controller. The 5v of the same segment splits to the power supply and to the level shifter. The second segment has only the data cable connected to the level shifter, its 5v and gnd are connected to the power supply only, could that be a problem?

**Comment by amigthea** [comment]
I have some other doubt on the controller too, is this the right one (ESP32 D1 MINI) ?

```
Detecting chip type... ESP32
Chip is ESP32-D0WD-V3 (revision v3.1)
```
